---
title: "HOmics preparing data"
author: "Lara Nonell"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  BiocStyle::html_document:
    number_sections: true
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

# Prepare data

integrate is the main function in the plackage. It needs several parameters:

- meth.data=GRSet
- pheno.cond.col="tissue.ch1"
- annot.gene.col="UCSC_RefGene_Name"
- annot.z.col="UCSC_RefGene_Group"
- annot.mult.sep=NULL
- z.matrix=NULL
- gene.list=genes.u
- cores=1
 
The package is prepared to analyze GenomicRatioSet (minfi) but also ExpressionSet (BioC), a standard class when downloading data using the GEOquery package.
In both cases it uses annotations 

parameters related to relative position of CpGs in a gene.

- 3'UTR
- TSS1500
- TSS200
- Body
- 5'UTR
 
## Analyzing a Methylation GenomicRatioSet

The package is prepared to analyze GenomicRatioSet (minfi) of methylated data in its default 

In this case default values are 

- annot.gene.col="UCSC_RefGene_Name"
- annot.z.cols=c("UCSC_RefGene_Group")
- annot.mult.sep=";"
- cores=1

annot.mult.sep defines the separator in the annot.gene.cols and annot.z.cols

## Other

## Response variable, phenotype or condition

pheno.cond.col should be a variable  in the annotation data  associated to meth.data
usually a column in pData(meth.data)

Response variable can be continuous or categorical. 

### Categorigal responses

When response variable is categorical, a hierarchical logistic model will be performed.

In such cases, some considerations need to be taken.

JAGS needs a vector with 0 and 1, so it will be converted if it is not the case
The rules to convert the values are the following.

if it is already numerical, will be converted to a vector of 0's and 1's using the module 2 function denoted in R with %%

```{r cond1}
a <- c(2,1,1,1,2)
y <- a %% 2
y
```

if it is a character variable, it will be converted to a numerical factor, to be interpreted by JAGS

Remember when converting to factor in R, it always assigns the levels in alphabetical order.
So for instance vector a with 4 character categories, will finally converted to a numerical vector of 0's and 1's this way:

```{r cond2}
a <- c("z","t","b","z","t","b","z")
(a.f <- as.factor(a))
levels(a.f)
(a.f.n <- as.numeric(a.f))
(cond <- a.f.n %% 2)
```

### Continuous responses

When response variable is continuous, a hierarchical linear model will be performed.

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

# Annotations

GEO data from Illumina methylation array 27k (GPL...) and 450K (GPL...) contain two fields UCSC_RefGene_Name and UCSC_RefGene_Group but EPIC (GPL21145) does not contain such information. Annotation files
