---
title: "HOmics methyaltion in genes integragion vignette"
author: "Lara Nonell"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  BiocStyle::html_document:
    number_sections: true
    toc_float: yes
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)

```

#load package

```{r HOmics}

library(HOmics)

```


# Prepare data

**HOmics.meth** function permits the analysis of hierarchical model considering the methylation beta-values of a microarray experiment. For other applications in the package refer to [HOmics vignette](HOmics.html)
**HOmics.meth** function needs several parameters:

- **meth.data**: GenomicRatioSet or  ExpressionSet containing methylation information as well as annotations on gene position
- **pheno.cond.col**: response variable, column of annotation that contains condition (one of the variable names contained in pheno data obtained with function pData())
- **pheno.covar.col covariates**: covariates, column of annotation that contains names of covariates to include in the model (one or more of the variable names contained in pheno data). Default = NULL
- **annot.gene.col**: column of annotation that contains gene names (same annotations as gene.list). Default = "UCSC_RefGene_Name"
- **annot.z.col**: prior information. Column of annotation that contains prior variable (for GenomicRatioSet functions getAnnotation() or fData() for ExpressionSet). Default = "UCSC_RefGene_Group"
- **annot.mult.sep**: separator for annot.gene.col and maybe annot.z.col multiple values. Default = ";"
- **z.matrix**: prior information. Alternative parameter to annot.z.col as matrix with rows the featureNames of meth.data and columns the Zvars can be given. Default = NULL
- **gene.list**: genes to analyze as a vector with symbols (HUGO)
- **cores**: cores in case of parallelization. Default = 1 (no parallelization)

 
The package is prepared to analyze GenomicRatioSet (minfi) but also ExpressionSet (BioC), a standard class when downloading data using the GEOquery package.
In both cases it uses annotations 

parameters related to relative position of CpGs in a gene.

- 3'UTR
- TSS1500
- TSS200
- Body
- 5'UTR
 

The function **HOmics.meth** is prepared to analyze GenomicRatioSet (minfi) or ExpressionSet (BioC) of methylated data.

In this case default values are 

- **annot.gene.col** = "UCSC_RefGene_Name"
- **annot.z.cols** = "UCSC_RefGene_Group"
- **annot.mult.sep** = ";"
- **cores** = 1

**annot.mult.sep** defines the separator in the annot.gene.cols and annot.z.cols


## Response variable, phenotype or condition

pheno.cond.col should be a variable  in the annotation data  associated to meth.data
usually a column in pData(meth.data)

Response variable can be continuous or categorical. 

### Categorigal responses

When response variable is categorical, a hierarchical logistic model will be performed.

In such cases, some considerations need to be taken.

JAGS needs a vector with 0 and 1, so it will be converted if it is not the case
The rules to convert the values are the following.

if it is already numerical, will be converted to a vector of 0's and 1's using the module 2 function denoted in R with %%

```{r cond1}

a <- c(2,1,1,1,2)
y <- a %% 2
y

```

if it is a character variable, it will be converted to a numerical factor, to be interpreted by JAGS

Remember when converting to factor in R, it always assigns the levels in alphabetical order.
So for instance vector a with 4 character categories, will finally converted to a numerical vector of 0's and 1's this way:

```{r cond2}

a <- c("z","t","b","z","t","b","z")
(a.f <- as.factor(a))
levels(a.f)
(a.f.n <- as.numeric(a.f))
(cond <- a.f.n %% 2)

```

### Continuous responses

When response variable is continuous, a hierarchical linear model will be performed.

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

## Annotations

GEO data from Illumina methylation array 27k (GPL8490) and 450K (GPL13534) contain two fields UCSC_RefGene_Name and UCSC_RefGene_Group but EPIC (GPL21145) does not contain such information. Annotation files for those objects should be added manually.

# Example 1: ExpressionSet from GEO 

ExpressionSet corresponds to GEO accession number GSE117929. Data was previously downloaded using package `r Biocpkg("GEOquery")` and accessible as data object in **HOmics**. `r Biocpkg("Biobase")` package is needed to manipulate ExpressionSet class objects.

GSE117929 contains a methylome-wide analysis of 37 samples of chronic HIV infected patients and healthy controls.

```{r load_data}

library(Biobase)

data("GSE117929", package="HOmics")

GSE117929

table(pData(GSE117929)$"diagnosis:ch1")

```


List of genes to model obtained from [PMC5988798](https://www.nature.com/articles/s41598-018-26894-4))


```{r model}

genes <- c("CCR5","CXCR4")

res.meth <- HOmics.meth(meth.data = GSE117929,
                   pheno.cond.col = "diagnosis:ch1",
                   pheno.covar.col = NULL,
                   annot.gene.col = "UCSC_RefGene_Name",
                   annot.z.col = "UCSC_RefGene_Group",
                   annot.mult.sep = ";",
                   z.matrix = NULL,
                   gene.list = genes,
                   cores = 1)

class(res.meth)

```

Object **res** contains a list with each analysis result performed for each genes with all related CpGs. Internally the function searches the annotation field UCSC_RefGene_Name for each gene and gets its gene relative position (3'UTR, TSS1500, TSS200, Body or
5'UTR) and generates a matrix of prior information ()

It is an object of class **HOmics**, which can be subsequently filtered to get significative features.

```{r model_res}

res.meth[[1]]

```

We filter the results of those CpGs in genes that have a significative positive or significative negative betas of the bayesian hierarchical model. In this example we filter at a significance level of 0.2 for demo purposes.

```{r filter_res}

res.f.pos <- getsignif(res.meth, param = "p.pos", threshold = 0.2, as.data.frame = T)
res.f.pos

res.f.neg <- getsignif(res.meth, param = "p.neg", threshold = 0.2, as.data.frame = T)
res.f.neg

```

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```